### DISCUSSION TITLE
How to Limiting generated code for Client API by  reuse  generated code to reduce code generation time?

### DISCUSSION THREAD

**Discussion by Rudy0530** [Netflix/dgs-framework#1451] [Q&A] [answered]
Thank you for providing a great open source for graphql. I have a question while using it.

My project uses graphqlcodegen-maven-plugin to generate graphql client code.

Programming language is  `java.`

Relevant dependencies are as follows.

```xml
        <dependency>
            <groupId>com.netflix.graphql.dgs.codegen</groupId>
            <artifactId>graphql-dgs-codegen-client-core</artifactId>
            <version>5.1.17</version>
        </dependency>
        <dependency>
            <groupId>com.netflix.graphql.dgs</groupId>
            <artifactId>graphql-dgs-client</artifactId>
            <version>5.5.1</version>
        </dependency>



        <plugin>
               <groupId>io.github.deweyjose</groupId>
                <artifactId>graphqlcodegen-maven-plugin</artifactId>
                <version>1.29</version>
                <configuration>
                    <schemaPaths>
                        <param>src/main/resources/schema</param>
                    </schemaPaths>
                    <generateClient>true</generateClient>
                    <maxProjectionDepth>4</maxProjectionDepth>
                    <javaGenerateAllConstructor>true</javaGenerateAllConstructor>
                    <packageName>packageName</packageName>
                    <subPackageNameClient>client</subPackageNameClient>
                    <subPackageNameDatafetchers>datafetchers</subPackageNameDatafetchers>
                    <subPackageNameTypes>types</subPackageNameTypes>
                    <typeMapping>
                        <Long>java.lang.Long</Long>
                    </typeMapping>
                </configuration>
        </plugin>
```

As we all know ,if schema  has a lot of cycles, it is not ideal to generate client APIs for the entire schema.

There have a few configuration parameters that help tune this so you can limit the generation of client API to only what is required.

```
generateJava {
    ...
    generateClient = true
    skipEntityQueries = true
    includeQueries = ["hello"]
    includeMutations = [""]
    includeSubscriptions = [""]
    maxProjectionDepth = 2
}
```

I use `maxProjectionDepth` to control the amount of graphql client generated.  When I increase the `maxProjectionDepth` parameter, code generation takes a long time, and even OOM occurs.

I found a lot of duplication in the generated client code, such as

```java

Device_Software_Organization_Device_OrganizationProjection.class
```

Whether we have parameters that can reuse nodes in the circle to reduce the amount of duplicate code generation, and reduce code generation time

Thank you very much for your reply.

**Comment by srinivasankavitha** [comment, marked as answer]
Hi,
We recently released an experimental feature for improving the generation
of the client API..it can be turned on using the `generateClientv2` flag.
We don't maintain the maven version of the plugin though it's available in
the gradle plugin, if you want to try that.

**Reply by Rudy0530** [reply]
Thank you very much for your reply. I'll try it. thank you.
